src/atc/transformations.py

```diff
@@ -197,3 +204,168 @@ def join_time_series_dataframes(
+            else:
+                filter_string += f" OR (a.{col} <> b.{col})"
+                filter_string += f" OR (a.{col} IS NULL AND b.{col} IS NOT NULL)"
+                filter_string += f" OR (a.{col} IS NOT NULL AND b.{col} IS NULL)"
+
+    df = (df.alias("a")
+          .join(df_target.alias("b"), on=join_cols, how="left")
+          .filter(filter_string)
+          .withColumn("is_new",
+                      F.when(f.col(f"{key_col}").isNull(), True)
+                      .otherwise(False))
+          .select("a.*", "is_new")
+          .cache()
+          )
+
+    # If merge is not required, the data can just be appended
+    merge_required = (len(df.filter(~ F.col("is_new")).take(1)) > 0)
+    df = df.drop("is_new")
+
+    if not merge_required:
+        return df.write \
+            .format(table_format) \
+            .mode("append") \
+            .save(target_table_name)
+
+    unique_id = uuid.uuid4().hex
+    temp_view_name = f"source_{unique_id}"
+    df.createOrReplaceTempView(temp_view_name)
+
+    non_join_cols = [col for col in df.columns if col not in join_cols]
+
+    merge_sql_statement = f"""
+        MERGE INTO {target_table_name} AS target
+        USING {temp_view_name} as source
+        ON
+        {" AND ".join(f"(source.{col} = target.{col})" for col in join_cols)}
+        WHEN MATCHED
+        THEN UPDATE -- update existing records
+        SET
+        {', '.join(f"target.{col} = source.{col}" for col in non_join_cols)}
+        WHEN NOT MATCHED
+        THEN INSERT -- insert new records
+        (
+        {', '.join(df.columns)}
+        )
+        VALUES
+        (
+        {', '.join(f"source.{col}" for col in df.columns)}
+        )
+        """
+    Spark.get().sql(merge_sql_statement)
+
+
+def concat_dfs(dfs: List[DataFrame]):
+    """
+       concat_dfs append dataframes on eachother and also append columns.
+            If columns has the exact same name, they will be converted to one column.
+
+       NB: The transformation does not perform dataframe joins.
+
+       :param p1: A list of spark dataframes
+
+       returns a dataframe
+
+       """
+
+    # Check that the list of dataframe and each dataframe is not Nonetypes
+    if dfs is None:
+        raise NoTableException(f"The table list of tables are None")
+
+    for i, df in enumerate(dfs):
+        if df is None:
+            raise NoTableException(f"The {i}th table dataframe in the list is None.")
+
+    all_cols = []
+    dfs_updated = dfs
+
+    # Get all columns across all tables
+    for df in dfs:
+        all_cols = all_cols + df.columns
+
+    # Sort the columns
+    all_cols = sorted(list(set(all_cols)))
+
+    # Create a list to save missing columns for each dataframe
+    missing = []
+
+    # For each dataframe: Get missing columns and add them to the dataframe
+    for i, df in enumerate(dfs):
+        missing.append([col for col in all_cols if col not in df.columns])
+
+        for col in missing[i]:
```

Comment:
- `missing` is never used outside the for loop. there is no reason to keep this state for later.
- the list comprehension that you construct adds nothing here.
- recommendation: do a simple set of nested for loops
```
    for i, df in enumerate(dfs):
         for col in all_cols:
            if col not in df.columns: 
                dfs_updated[i] = dfs_updated[i].withColumn(col, F.lit(None))
```
- this code is much cleaner and would have been easier for me to read.